src/tickets/types.ts

```diff
@@ -0,0 +1,138 @@
+import {FirebaseFirestoreTypes} from '@react-native-firebase/firestore';
+
+enum TravelRightStatus {
+  UNSPECIFIED = 0,
+  RESERVED = 1,
+  ORDERED = 2,
+  PAID_FOR = 3,
+  UNUSED = 4,
+  ACTIVATED = 5,
+  PARTIALLY_USED = 6,
+  USED = 7,
+  ARCHIVED = 8,
+  OTHER = 9,
+  CANCELLED = 20,
+  REFUNDED = 21,
+}
+
+export type TravelRight = {
+  id: string;
+  status: TravelRightStatus;
+  type:
+    | 'PreActivatedSingleTicket'
+    | 'PreActivatedPeriodTicket'
+    | 'UnknownTicket';
+};
+
+export type Timestamp = FirebaseFirestoreTypes.Timestamp;
+
+export type PreactivatedTicket = TravelRight & {
+  fareProductRef: string;
+  startDateTime: Timestamp;
+  endDateTime: Timestamp;
+  usageValidityPeriodRef: string;
+  userProfileRef: string;
+  authorityRefs: string[];
+  tariffZoneRefs: string[];
+};
+
+export type PreactivatedSingleTicket = PreactivatedTicket & {
+  type: 'PreActivatedSingleTicket';
```

Comment:
Doesn't it need to be `Omit<PreactivatedTicket, 'type'> & { type: ... }`?

Reply:
No the `type` will be forced to `PreActivatedSingleTicket` here. If I try to set anything else it will not compile.
![image](https://user-images.githubusercontent.com/4932625/107917706-a6c2d380-6f68-11eb-95fc-9d6f64abeea6.png)

Reply:
But might be another reason to use `Omit` that I'm not seeing instantly?

Reply:
Ok. It may just be Intellij which does not handle this optimally then. Without the `Omit` it resolves usage of `type` to have two possible declarations, both the one from `PreactivatedTicket` and the one from `PeriodTicket`. Although it I see that the typescript compiler gives no error.

Reply:
It doesn't give an error for you? What's the code you're trying? For the above code I'm getting the following when trying the compiler:
```bash
∟ yarn tsc
yarn run v1.22.10
$ /Users/christianbrevik/Dev/AtB/mittatb-app/node_modules/.bin/tsc
src/tickets/types.ts:48:3 - error TS2322: Type '"UnknownTicket"' is not assignable to type '"PreActivatedPeriodTicket"'.

48   type: 'UnknownTicket',
     ~~~~

  src/tickets/types.ts:21:3
    21   type:
         ~~~~
    The expected type comes from property 'type' which is declared here on type 'PeriodTicket'
```

Reply:
Yeah, I get the error, so no problem there. It is when reading the `type` it is not handled optimally in Intellij, as it interprets that `type` has two definitions. But no biggie, the typescript compiler is not confused!

<img width="695" alt="Screenshot 2021-02-15 at 09 58 05" src="https://user-images.githubusercontent.com/675421/107925181-5f424480-6f74-11eb-8340-b3d95815d646.png">
<img width="603" alt="Screenshot 2021-02-15 at 09 58 24" src="https://user-images.githubusercontent.com/675421/107925178-5ea9ae00-6f74-11eb-8304-e261d5eca727.png">